File: 0019-remove-nth-node-from-end-of-list/0019-remove-nth-node-from-end-of-list.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def removeNthFromEnd(self, head: Optional[ListNode], n: int) -> Optional[ListNode]:
        # Logic -- Barick
        # take a fast and a slow pointer
        # if create a distance of N+1 between them
        # when the fast pointer reaches last node,
        # we'll know that need to remove the slow+1 node
        slow = head
        fast = head

        # both fast and slow is at 1 i.e. head
        # move fast to another n positions ahead
        # so the distance becomes N+1
        for i in range(n):
            if fast is None:
                break
            else:
                fast = fast.next

        # Out of bounds check
        if fast is None:
            # it means we need to remove the head itself
            newHead = head.next 
            head.next = None
            return newHead

        # keep fowarding both the pointers,
        # until fast moves to last node
        # because at this point slow will be at (n+1) from last
        while fast.next != None:
            slow = slow.next
            fast = fast.next

        # we need to remove slow+1 th node
        # so, connect slow with its next.next node
        nodeToRemove = slow.next
        slow.next = slow.next.next
        nodeToRemove.next = None

        return head

File: 0019-remove-nth-node-from-end-of-list/0019-remove-nth-node-from-end-of-list.js
/**
 * Definition for singly-linked list.
 * function ListNode(val, next) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.next = (next===undefined ? null : next)
 * }
 */
/**
 * @param {ListNode} head
 * @param {number} n
 * @return {ListNode}
 */
var removeNthFromEnd = function(head, n) {
    // let's take 2 pointers at p1 and pn,
    // they are in relative position of nth to each other
    // let's keep forwarding them together, their distance remaining const
    // when pn reaches end, we need to delete the p1, 
    // (and to do that we need to keep track of p0 also) 
    let p1 = head, p0 = null, pn = p1;

    // forward pn to (n-1) steps
    for(let i=1; i < n; i++) pn = pn.next;

    // at this point they are nth node from each other
    while(pn) {
        // if pn is last node
        // p1 needs to be removed
        if(!pn.next) {
            // Careful if it's the head
            // at this point p0 is null, 
            // as there is no node before head
            if(p1 === head) head = head.next;
            else p0.next = p1.next; // DONE!!!
        }

        // right shift each
        pn = pn.next;
        p0 = p1;
        p1 = p1.next;
    }

    return head;
};